portfolio/src/main/java/com/google/sps/servlets/MemeHandler.java

```diff
@@ -0,0 +1,126 @@
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("Meme").addSort("timeStamp", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+    // List<Entity> resultsList = results.asList(FetchOptions.Builder.withLimit(maxMemes));
+
+    List<Meme> memes = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+        long id = entity.getKey().getId();
+        String author = (String) entity.getProperty("author");
+        String url = (String) entity.getProperty("url");
+        String description = (String) entity.getProperty("desc");
+        String timestamp = (String) entity.getProperty("timeStamp");
+
+        Meme memeObject = new Meme(id, author, url, description, timestamp);
+        memes.add(memeObject);
+    }
+
+    response.setContentType("application/json;");
+    String json = gson.toJson(memes);
+    response.getWriter().println(json);
+    }
+
+
+    /** Receive a meme and upload its data to the datastore. */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String description = request.getParameter("message");
+
+    String imageUrl = getUploadedFileUrl(request, "file");
+
+    Entity memeEntity = new Entity("Meme");
+    memeEntity.setProperty("author", "anonymous");
+    memeEntity.setProperty("url", imageUrl);
+    memeEntity.setProperty("desc", description);
+    memeEntity.setProperty("timeStamp", "11/22/11");
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(memeEntity);
+    response.sendRedirect("/meme.html");
+  }
+
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // We could check the validity of the file here, e.g. to make sure it's an image file
+    // https://stackoverflow.com/q/10779564/873165
```

Comment:
should this be a todo?

Reply:
sorry i forgot to remove all of those before i pushed, ignore for now